File: template/app/src/main/java/com/xmartlabs/template/helper/GeneralFlowableSubscriber.java
package com.xmartlabs.template.helper;

import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;

import com.xmartlabs.template.ui.common.TemplateView;

import org.reactivestreams.Subscription;

import java.lang.ref.WeakReference;

import io.reactivex.FlowableSubscriber;

public class GeneralFlowableSubscriber<T> implements FlowableSubscriber<T> {
  @NonNull
  private WeakReference<TemplateView> viewReference;

  public GeneralFlowableSubscriber() {
    this(null);
  }

  public GeneralFlowableSubscriber(@Nullable TemplateView templateView) {
    viewReference = new WeakReference<>(templateView);
  }

  @Override
  public void onSubscribe(@NonNull Subscription subscription) {
    subscription.request(getMaxNumberOfElements());
  }

  @Override
  public void onNext(T t) {}

  @Override
  public void onError(Throwable throwable) {
    TemplateView view = viewReference.get();
    if (alertOnError(throwable) && view != null && view.isViewAlive()) {
      view.showError(throwable, getErrorMessage(throwable));
    }
  }

  @Override
  public void onComplete() {}

  @Nullable
  @StringRes
  protected Integer getErrorMessage(Throwable throwable) {
    return null;
  }

  @CheckResult
  protected boolean alertOnError(Throwable throwable) {
    return true;
  }

  protected long getMaxNumberOfElements() {
    return Long.MAX_VALUE;
  }
}

File: template/app/src/main/java/com/xmartlabs/template/helper/GeneralFlowableSubscriber.kt
package com.xmartlabs.template.helper

import android.support.annotation.CheckResult
import android.support.annotation.StringRes
import com.xmartlabs.template.ui.common.TemplateView
import io.reactivex.FlowableSubscriber
import org.reactivestreams.Subscription
import java.lang.ref.WeakReference

open class GeneralFlowableSubscriber<T> constructor(templateView: TemplateView? = null) : FlowableSubscriber<T> {
  protected val maxNumberOfElements = java.lang.Long.MAX_VALUE
  private val viewReference: WeakReference<TemplateView> = WeakReference<TemplateView>(templateView)

  override fun onSubscribe(subscription: Subscription) = subscription.request(maxNumberOfElements)

  override fun onNext(t: T) { }

  override fun onError(throwable: Throwable) {
    val view = viewReference.get()
    if (alertOnError(throwable) && view != null && view.isViewAlive) {
      view.showError(throwable, getErrorMessage(throwable))
    }
  }

  override fun onComplete() { }

  @StringRes
  protected open fun getErrorMessage(throwable: Throwable): Int? = null

  @CheckResult
  protected open fun alertOnError(throwable: Throwable) = true
}
